esphome/components/st7920/st7920.cpp

```diff
@@ -0,0 +1,148 @@
+#include "st7920.h"
+#include "esphome/core/log.h"
+#include "esphome/core/application.h"
+#include "esphome/components/display/display_buffer.h"
+
+namespace esphome {
+namespace st7920 {
+
+static const char *const TAG = "st7920";
+
+// ST7920 COMMANDS
+static const uint8_t LCD_DATA = 0xFA;
+static const uint8_t LCD_COMMAND = 0xF8;
+static const uint8_t LCD_CLS = 0x01;
+static const uint8_t LCD_HOME = 0x02;
+static const uint8_t LCD_ADDRINC = 0x06;
+static const uint8_t LCD_DISPLAYON = 0x0C;
+static const uint8_t LCD_DISPLAYOFF = 0x08;
+static const uint8_t LCD_CURSORON = 0x0E;
+static const uint8_t LCD_CURSORBLINK = 0x0F;
+static const uint8_t LCD_BASIC = 0x30;
+static const uint8_t LCD_GFXMODE = 0x36;
+static const uint8_t LCD_EXTEND = 0x34;
+static const uint8_t LCD_TXTMODE = 0x34;
+static const uint8_t LCD_STANDBY = 0x01;
+static const uint8_t LCD_SCROLL = 0x03;
+static const uint8_t LCD_SCROLLADDR = 0x40;
+static const uint8_t LCD_ADDR = 0x80;
+static const uint8_t LCD_LINE0 = 0x80;
+static const uint8_t LCD_LINE1 = 0x90;
+static const uint8_t LCD_LINE2 = 0x88;
+static const uint8_t LCD_LINE3 = 0x98;
+
+void ST7920::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up ST7920...");
+  this->dump_config();
+  this->spi_setup();
+  this->init_internal_(this->get_buffer_length_());
+  display_init_();
+}
+
+void ST7920::command_(uint8_t value) {
+  this->enable();
+  this->send_(LCD_COMMAND, value);
+  this->disable();
+}
+
+void ST7920::data_(uint8_t value) {
+  this->enable();
+  this->send_(LCD_DATA, value);
+  this->disable();
+}
+
+void ST7920::send_(uint8_t type, uint8_t value) {
+  this->write_byte(type);
+  this->write_byte(value & 0xF0);
+  this->write_byte(value << 4);
+}
+
+void ST7920::goto_xy_(uint16_t x, uint16_t y) {
+  if (y >= 32 && y < 64) {
+    y -= 32;
+    x += 8;
+  } else if (y >= 64 && y < 64 + 32) {
+    y -= 32;
+    x += 0;
+  } else if (y >= 64 + 32 && y < 64 + 64) {
+    y -= 64;
+    x += 8;
+  }
+  this->command_(LCD_ADDR | y);  // 6-bit (0..63)
+  this->command_(LCD_ADDR | x);  // 4-bit (0..15)
+}
+
+void HOT ST7920::write_display_data() {
+  byte i, j, b;
```

Comment:
`byte` is an Arduino-ism, we prefer standard C++.
```suggestion
  uint8_t i, j, b;
```